packages/react-pages/src/node/dynamic-modules/PageStrategy.ts

```diff
@@ -0,0 +1,282 @@
+import * as path from 'path'
+import { EventEmitter } from 'events'
+import chokidar, { FSWatcher } from 'chokidar'
+import { debounce } from 'mini-debounce'
+import { PagesData } from './pages'
+import {
+  extractStaticData,
+  defaultPageLoader,
+  defaultPageFinder,
+} from './utils'
+
+export class PageStrategy extends EventEmitter {
+  readonly getPages: () => Promise<Readonly<PagesData>>
+  readonly close: () => void
+
+  constructor(
+    pagesDir: string,
+    findPages: FindPages = defaultPageFinder,
+    loadPageData: LoadPageData = defaultPageLoader
+  ) {
+    super()
+
+    const pageCache: PagesData = {}
+    const fileCache: FileCache = {}
+    const watchers = new Set<FSWatcher>()
+
+    // The "promise" event is for HMR updates to the pages map.
+    const emitPromise = debounce(() => {
+      this.emit('promise', pagesPromise)
+    }, 100)
+
+    // The "change" event is for HMR updates to each page.
+    const changedPages = new Set<string>()
+    const emitChange = debounce(() => {
+      this.emit('change', Array.from(changedPages))
+      changedPages.clear()
+    }, 100)
+
+    const setPageData = (pageId: string, pageData: Partial<PageData>) => {
+      if (!pageId.startsWith('/'))
+        throw Error(
+          `setPageData error: pageId should start with "/", but got "${pageId}"`
+        )
+
+      const existingPage = pageCache[pageId]
+      const page =
+        existingPage ||
+        (pageCache[pageId] = {
+          data: {},
+          staticData: {},
+        })
+
+      const key = pageData.key ?? 'main'
+      if (pageData.dataPath) {
+        if (page.data[key])
+          throw Error(
+            `setPageData conflict: data with key "${key}" already exists`
+          )
+        page.data[key] = pageData.dataPath
+      }
+      if (pageData.staticData) {
+        if (page.staticData[key])
+          throw Error(
+            `setPageData conflict: staticData with key "${key}" already exists`
+          )
+        page.staticData[key] = pageData.staticData
+      }
+
+      if (existingPage) {
+        changedPages.add(pageId)
+        emitChange()
+      }
+    }
+
+    const removePageData = (pageId: string, key?: string) => {
+      const page = pageCache[pageId]
+      if (page) {
+        if (key) {
+          delete page.data[key]
+          delete page.staticData[key]
+          changedPages.add(pageId)
+          emitChange()
+        } else {
+          delete pageCache[pageId]
+        }
+      }
+    }
+
+    const watchFiles = (
+      baseDir: string,
+      arg2?: string | string[] | FileHandler,
+      arg3?: FileHandler
+    ) => {
+      let globs: string[]
+      let handler: (
+        file: File,
+        type: string
+      ) => PageData | Promise<PageData> | void
+
+      if (typeof arg2 === 'function') {
+        globs = ['**/*']
+        handler = arg2
+      } else {
+        globs = typeof arg2 === 'string' ? [arg2] : arg2!
+        handler = arg3 || defaultPageLoader
```

Comment:
What if both arg2 and arg3 is undefined? We should throw error for this, right?

Reply:
Will do 👍

Reply:
Technically, that scenario is guarded by TypeScript. Do you still want a runtime check?

Reply:
I don't think this is guarded by ts yet. Because current `WatchFilesHelper` have this signature: `(baseDir: string, handler?: FileHandler): void`